### PR TITLE
Always set OIDC tenant id and document how to access it in Hibernate and Mongo resolvers

### LINKED ISSUE #44168: Unable to read RoutingContext from CustomTenantResolver
### Describe the bug

I am trying to read the tenant from RoutingContext, to resolve the database connection based on tenantId, but getting this error

```
jakarta.enterprise.inject.IllegalProductException: Normal scoped producer method may not return null: io.quarkus.vertx.http.runtime.CurrentVertxRequest.getCurrent()       
        at io.quarkus.vertx.http.runtime.CurrentVertxRequest_ProducerMethod_getCurrent_6dc23d16d53ba5c34e1e7b6f54290fd7b9aebd76_Bean.doCreate(Unknown Source)
        at io.quarkus.vertx.http.runtime.CurrentVertxRequest_ProducerMethod_getCurrent_6dc23d16d53ba5c34e1e7b6f54290fd7b9aebd76_Bean.create(Unknown Source)
        at io.quarkus.vertx.http.runtime.CurrentVertxRequest_ProducerMethod_getCurrent_6dc23d16d53ba5c34e1e7b6f54290fd7b9aebd76_Bean.create(Unknown Source)
        at io.quarkus.arc.impl.RequestContext.getIfActive(RequestContext.java:74)
        at io.quarkus.arc.impl.ClientProxies.getDelegate(ClientProxies.java:30)
        at io.vertx.ext.web.CurrentVertxRequest_ProducerMethod_getCurrent_6dc23d16d53ba5c34e1e7b6f54290fd7b9aebd76_ClientProxy.arc$delegate(Unknown Source)
        at io.vertx.ext.web.CurrentVertxRequest_ProducerMethod_getCurrent_6dc23d16d53ba5c34e1e7b6f54290fd7b9aebd76_ClientProxy.get(Unknown Source)
        at org.acme.project.solver.CustomTenantResolver.resolveTenantId(CustomTenantResolver.java:23)
        at org.acme.project.solver.CustomTenantResolver_ClientProxy.resolveTenantId(Unknown Source)
        at io.quarkus.hibernate.orm.runtime.tenant.HibernateCurrentTenantIdentifierResolver.resolveCurrentTenantIdentifier(HibernateCurrentTenantIdentifierResolver.java:37)
        at org.hibernate.internal.SessionFactoryImpl$SessionBuilderImpl.<init>(SessionFactoryImpl.java:1211)
        at org.hibernate.internal.SessionFactoryImpl.withOptions(SessionFactoryImpl.java:607)
        at org.hibernate.internal.SessionFactoryImpl.openSession(SessionFactoryImpl.java:583)
        at org.hibernate.internal.SessionFactoryImpl.openSession(SessionFactoryImpl.java:154)
        at io.quarkus.hibernate.orm.runtime.RequestScopedSessionHolder.lambda$getOrCreateSession$0(RequestScopedSessionHolder.java:21)
        at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1220)
        at io.quarkus.hibernate.orm.runtime.RequestScopedSessionHolder.getOrCreateSession(RequestScopedSessionHolder.java:21)
        at io.quarkus.hibernate.orm.runtime.RequestScopedSessionHolder_ClientProxy.getOrCreateSession(Unknown Source)
        at io.quarkus.hibernate.orm.runtime.session.TransactionScopedSession.acquireSession(TransactionScopedSession.java:104)
        at io.quarkus.hibernate.orm.runtime.session.TransactionScopedSession.createQuery(TransactionScopedSession.java:366)



        at org.hibernate.engine.spi.SessionLazyDelegator.createQuery(SessionLazyDelegator.java:548)
        at org.hibernate.engine.spi.SessionLazyDelegator.createQuery(SessionLazyDelegator.java:66)
        at org.hibernate.Session_3a974b6a18ac399f675913d732c105426414d370_Synthetic_ClientProxy.createQuery(Unknown Source)
        at io.quarkus.hibernate.orm.panache.common.runtime.CommonPanacheQueryImpl.createBaseQuery(CommonPanacheQueryImpl.java:353)
        at io.quarkus.hibernate.orm.panache.common.runtime.CommonPanacheQueryImpl.createQuery(CommonPanacheQueryImpl.java:315)
        at io.quarkus.hibernate.orm.panache.common.runtime.CommonPanacheQueryImpl.list(CommonPanacheQueryImpl.java:266)
        at io.quarkus.hibernate.orm.panache.runtime.PanacheQueryImpl.list(PanacheQueryImpl.java:149)
        at org.acme.project.solver.CacheResolver.refreshCache(CacheResolver.java:22)
        at org.acme.project.solver.AppLifecycleBean.onStart(AppLifecycleBean.java:16)
        at org.acme.project.solver.AppLifecycleBean_Observer_onStart_a58341e9b5ff8692203684dfc1ea68ef736fee23.notify(Unknown Source)
        at io.quarkus.arc.impl.EventImpl$Notifier.notifyObservers(EventImpl.java:346)
        at io.quarkus.arc.impl.EventImpl$Notifier.notify(EventImpl.java:328)
        at io.quarkus.arc.impl.EventImpl.fire(EventImpl.java:82)
        at io.quarkus.arc.runtime.ArcRecorder.fireLifecycleEvent(ArcRecorder.java:155)
        at io.quarkus.arc.runtime.ArcRecorder.handleLifecycleEvents(ArcRecorder.java:106)
        at io.quarkus.deployment.steps.LifecycleEventsBuildStep$startupEvent1144526294.deploy_0(Unknown Source)
        at io.quarkus.deployment.steps.LifecycleEventsBuildStep$startupEvent1144526294.deploy(Unknown Source)
        ... 13 more 
```

### Expected behavior

Read RoutingContext keys from CustomTenantResolver

### Actual behavior

Getting jakarta.enterprise.inject.IllegalProductException: Normal scoped producer method may not return null: io.quarkus.vertx.http.runtime.CurrentVertxRequest.getCurrent()       

### How to Reproduce?

```
import io.quarkus.hibernate.orm.PersistenceUnitExtension;
import io.quarkus.hibernate.orm.runtime.tenant.TenantResolver;
import io.vertx.ext.web.RoutingContext;
import jakarta.enterprise.context.RequestScoped;
import jakarta.inject.Inject;

@PersistenceUnitExtension
@RequestScoped
public class CustomTenantResolver implements TenantResolver {

    @Inject
    RoutingContext context;

    @Override
    public String getDefaultTenantId() {
        return "default";
    }

    @Override
    public String resolveTenantId() {
        return context.get("tenantId");
    }
}
```

Setting the tenant Id in my OIDC resolver

```
import org.eclipse.microprofile.config.ConfigProvider;
import io.smallrye.mutiny.Uni;
import io.quarkus.oidc.OidcRequestContext;
import io.quarkus.oidc.OidcTenantConfig;
import io.quarkus.oidc.TenantConfigResolver;
import io.vertx.ext.web.RoutingContext;
import io.vertx.mutiny.core.Vertx;

@ApplicationScoped
public class CustomTenantOIDCResolver implements TenantConfigResolver {

    @Inject
    Vertx vertx;

    @Override
    public Uni<OidcTenantConfig> resolve(RoutingContext context, OidcRequestContext<OidcTenantConfig> requestContext) {
        String path = context.request().path();
        String[] parts = path.split("/");

        if (parts.length == 0) {
            // Resolve to default tenant configuration
            return null;
        }

        if (parts.length > 1) {
            return Uni.createFrom().item(createTenantConfig(parts[1], context));
        }

        // Resolve to default tenant configuration
        return null;
    }

    private Supplier<OidcTenantConfig> createTenantConfig(String clientId, RoutingContext context) {
        final OidcTenantConfig config = new OidcTenantConfig();
        List<Tenant> tenants = CacheResolver.tenants;
        tenants.stream().filter(t -> t.getFrontendClientId().equals(clientId)).findFirst().ifPresent(t -> {
            config.setTenantId(clientId);
            String authServer = ConfigProvider.getConfig().getValue("quarkus.oidc.auth-server-url", String.class);
            String baseAuthServer = authServer.substring(0, authServer.lastIndexOf("/"));
            config.setAuthServerUrl(baseAuthServer + "/" + clientId);
            config.setClientId(t.getBackendClientId());
            OidcTenantConfig.Credentials credentials = new OidcTenantConfig.Credentials();
            credentials.setSecret(t.getSecretId());
            config.setCredentials(credentials);
            context.put("clientId", config.getClientId().get());
            context.put("tenantId", config.getTenantId().get());
        });
        return () -> config;
    }

}

```

My database connection resolver

```
import io.agroal.api.AgroalDataSource;
import io.agroal.api.configuration.supplier.AgroalConnectionFactoryConfigurationSupplier;
import io.agroal.api.configuration.supplier.AgroalConnectionPoolConfigurationSupplier;
import io.agroal.api.configuration.supplier.AgroalDataSourceConfigurationSupplier;
import io.agroal.api.security.NamePrincipal;
import io.agroal.api.security.SimplePassword;
import io.quarkus.hibernate.orm.PersistenceUnitExtension;
import io.quarkus.hibernate.orm.runtime.customized.QuarkusConnectionProvider;
import io.quarkus.hibernate.orm.runtime.tenant.TenantConnectionResolver;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.enterprise.context.control.ActivateRequestContext;
import org.eclipse.microprofile.config.inject.ConfigProperty;
import org.hibernate.engine.jdbc.connections.spi.ConnectionProvider;

import java.sql.SQLException;
import java.util.HashMap;
import java.util.Map;

@ApplicationScoped
@ActivateRequestContext
@PersistenceUnitExtension
public class DataSourceTenantConnectionResolver implements TenantConnectionResolver {
    // @Inject
    // Instance<RoutingContext> contextInstance;

    @ConfigProperty(name = "quarkus.datasource.jdbc.url")
    String url;
    @ConfigProperty(name = "quarkus.datasource.username")
    String username;
    @ConfigProperty(name = "quarkus.datasource.password")
    String password;
    private final Map<String, ConnectionProvider> map = new HashMap<>();

    @Override
    public ConnectionProvider resolve(String tenantId) {
        System.out.println("resolve ConnectionProvider " + tenantId);
        if (map.containsKey(tenantId)) {
            return map.get(tenantId);
        }
        ConnectionProvider provider;
        provider = new QuarkusConnectionProvider(createDataSource(url, "mydb", username, password));
        map.put(tenantId, provider);
        return provider;
    }

    private AgroalDataSource createDataSource(
            String url, String database, String username, String password) {
        try {
            AgroalDataSourceConfigurationSupplier configurationSupplier = new AgroalDataSourceConfigurationSupplier();
            AgroalConnectionPoolConfigurationSupplier connectionPoolConfig = configurationSupplier
                    .connectionPoolConfiguration();
            connectionPoolConfig.maxSize(1000);

            AgroalConnectionFactoryConfigurationSupplier connectionFactoryConfig = connectionPoolConfig
                    .connectionFactoryConfiguration();
            connectionFactoryConfig.jdbcUrl(url);
            connectionFactoryConfig.principal(new NamePrincipal(username));
            connectionFactoryConfig.credential(new SimplePassword(password));

            return AgroalDataSource.from(configurationSupplier.get());
        } catch (SQLException | RuntimeException e) {
            throw new IllegalStateException("Exception while creating datasource for " + url, e);
        }
    }
}
```

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

I tried every solution on Internet with no luck

### PR DESCRIPTION
**Pull request by michalvavrik**
- closes: https://github.com/quarkusio/quarkus/issues/44168

### CODE REVIEW

**Comment by michalvavrik** [comment]
\cc @yrodiere in case you want to review, thanks

**Review by sberyozkin — COMMENTED** [review]
Michal, thanks,  I'm pretty sure we have a note about it in one of the Quarkus Hibernate docs.

As far as setting tenant id is concerned, OIDC sets it itself when the session cookie exists, and @TenantId resolver sets it too.
We should probably have it set in DefaultTenantConfigResolver if it is not already set.
IMHO it will be better

**Comment by michalvavrik** [comment]
> Michal, thanks, I'm pretty sure we have a note about it in one of the Quarkus Hibernate docs.

I can see notes in Hibernate docs and Mongo docs, however CDI request context activation in auth should concern authentication. I am just completing here incomplete existing documentation. Are you sure you want to duplicate this inside Hibernate (and Mongo) about proactive authentication?

> As far as setting tenant id is concerned, OIDC sets it itself when the session cookie exists, and @TenantID resolver sets it too. We should probably have it set in DefaultTenantConfigResolver if it is not already set. IMHO it will be better

I don't mind, but then maybe drop the existing example? Please look a few lines above the text I am adding.

Anyway that is really not a point of change here. Let me try to summarize what I try to document here, maybe I put it wrong into the words? I try to say:

- you want to use `RoutingContext`, so if you use `@ActivateRequestContext` and the CDI request context had not been active already, you will get CDI request context without `RoutingContext` (because how exactly woudl it get there without us setting it)
- if you call db from scheduler, you cannot inject `RoutingContext`, please inject `CurrentVertxRequest` instead

**Comment by michalvavrik** [comment]
@sberyozkin if you meant that I am putting this information into a wrong place or that I should duplicate it into Hibernate and Mongo guides, np, I am just not sure if I understood you, so I'll wait for a response.

**Comment by sberyozkin** [comment]
Sorry for the confusion @michalvavrik, indeed, what I thought was available in the Hibernate docs is indeed sitting just above the Hibernate TenantResolver example you added in this PR :-), where I show the custom OIDC resolver propagating tenant id via a routing context attribute.

But what your PR highlights is that code example I added earlier is outdated, the OIDC multi-tenancy docs should not ask users to propagate OIDC tenant id like I did in that example, instead, Quarkus OIDC should do it itself by setting a `tenant-id` attribute (when it is already set), and then the example you add would show that a `tenant-id` attribute should be retrieved, not `tenantId` as shown in my outdated example...

**Comment by michalvavrik** [comment]
@sberyozkin what I try to show is important for users if they want to get the OIDC tenant id inside of the Hibernate tenant resolver. Should I rewrite that and move this information to https://quarkus.io/guides/hibernate-orm#writing-the-application ? And maybe drop that original example with link to the Hibernate one?

**Comment by sberyozkin** [comment]
@michalvavrik I support this PR, it just depends on the outdated OIDC resolution example, which sets a custom `tenantId` attribute   while Quarkus OIDC already manages it with the tenant-id attribute. 
So what I propose is to make sure `tenant-id` is always set, remove the outdated oidc resolver example, and update your PR to show that `tenant-id` is read

**Review by yrodiere — DISMISSED** [review]

**Comment by michalvavrik** [comment]
@sberyozkin regarding the scheduler scenario (when there is no active CDI request context), I think it can be easily handled by changing `RoutingContext` injection point inside of the `OidcSessionImpl` to the `CurrentVertxRequest`. I'll follow up in a separate PR when I have more time to write a test for that. Cheers.

**Comment by sberyozkin** [comment]
@michalvavrik 

>regarding the scheduler scenario (when there is no active CDI request context), I think it can be easily handled by changing RoutingContext injection point inside of the OidcSessionImpl to the CurrentVertxRequest

I'm sorry, I'm not sure why the scheduler scenario is introduced, I certainly did not raise it. The example with `OidcSession` does not work with bearer access tokens and will confuse users. IMHO your original proposed Doc update was better, and there I only asked to align it with the current status quo, where Quarkus OIDC already sets `tenant-id`, and Hibernate resolvers checking it, as opposed to us asking users to set some other custom `tenantId` attribute, duplicating what Quarkus OIDC already does...

When I said we need to make sure tenant id is always set, I did not mean scheduled scenarios. I'm only not 100% `tenant-id` is set by OIDC `DefaultTenantResolverBean` in all the cases, which is what I meant to be double checked.

Sorry for the confusion, IMHO we need to avoid referring to OidcSession

File: docs/src/main/asciidoc/hibernate-orm.adoc
Comment (sberyozkin) [inline docs/src/main/asciidoc/hibernate-orm.adoc:1163]: It does not work with bearer access tokens

**Comment by michalvavrik** [comment]
> I'm sorry, I'm not sure why the scheduler scenario is introduced, I certainly did not raise it. 

User tried to access `RoutingContext` for DB queries from scheduler: https://github.com/quarkusio/quarkus/issues/44168

> The example with OidcSession does not work with bearer access tokens and will confuse users. IMHO your original proposed Doc update was better, and there I only asked to align it with the current status quo, where Quarkus OIDC already sets tenant-id

Do you say that `tenant-id` is set by OIDC and user should get it? Why should users work manually with some attribute on `RoutingContext` when OIDC retrieves it completely same?

https://github.com/quarkusio/quarkus/blob/1e7e87423c4071684dd0abba76c22c08f8265ed9/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcSessionImpl.java#L33

> When I said we need to make sure tenant id is always set, I did not mean scheduled scenarios. I'm only not 100% sure tenant-id is set by OIDC DefaultTenantResolverBean in all the cases, which is what I meant to be double checked.

It isn't, and I did double check. However it didn't make sense to me, there is `OIdcSession` where user can get tenant id and it doesn't work? I thought I was missing something.

File: docs/src/main/asciidoc/hibernate-orm.adoc
Comment (michalvavrik) [reply docs/src/main/asciidoc/hibernate-orm.adoc:1163]: I am probably missing something, I 'll wait for answer below before I will address this comment. Thanks

**Comment by michalvavrik** [comment]
> Sorry for the confusion, IMHO we need to avoid referring to OidcSession

Ok.

**Comment by sberyozkin** [comment]
Hi Michal, I was just thinking, so https://github.com/quarkusio/quarkus/blob/main/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcSessionImpl.java#L33 works, but it is a wrong concept on the bearer access token path so IMHO we should indeed avoid recommending it as a general tenant id check mechanism... I'll probably need to tighten its implementation to throw illegal access exception when some of its methods are called...

In any case sorry if I was not clear, if you'd like, I can look at your original PR update and tune it a bit as proposed to save you some time...

**Comment by michalvavrik** [comment]
I misunderstood `OidcSession`, I thought it was nice CDI bean that allows to avoid dealing with `RoutingContext` manually, but it is only meant for things related to this OIDC session https://openid.net/specs/openid-connect-session-1_0.html#CreatingUpdatingSessions. Sorry, I'll deal with fixes later today.

**Comment by michalvavrik** [comment]
So I have done whatever I understood from @sberyozkin comment. Sorry, but this may take few iterations. Also I don't have time to run all related tests locally as I have to go (run just few modules), so I'll just trigger CI and look at potential failures.

**Comment by michalvavrik** [comment]
@yrodiere I have requested review from you because I don't have permissions to just dismiss your review. I don't think it makes sense for you to review ATM because this might take a while. Thanks that you looked before.

**Review by yrodiere — COMMENTED** [review]
Commenting to finish the dismissing of my review :)
Agreed, it's more about OIDC than Hibernate ORM in the end, so I'm not much help here.

**Comment by sberyozkin** [comment]
Thanks @yrodiere

**Review by sberyozkin — APPROVED** [review]
Thanks @michalvavrik

**Pull request merged**
